File: sample_apps_reinvent2021/js/multi-value-attributes-example.js
const constants = require('./constants');

async function writeRecordsWithMultiMeasureValueSingleRecord() {
    console.log("Writing records with Multi Measure Value single record");
    const currentTime = Date.now().toString(); // Unix time in milliseconds

    const dimensions = [
        {'Name': 'region', 'Value': 'us-east-1'},
        {'Name': 'az', 'Value': 'az1'},
        {'Name': 'hostname', 'Value': 'host1'}
    ];

    const commonAttributes = {
        'Dimensions': dimensions,
        'Time': currentTime.toString()
    };

    const cpuUtilization = {
        'Name': 'cpu_utilization',
        'Value': '13.5',
        'Type': 'DOUBLE',
    };

    const memoryUtilization = {
        'Name': 'memory_utilization',
        'Value': '40',
        'Type': 'DOUBLE',
    };

    const computationalRecord = {
        'MeasureName': 'cpu_memory',
        'MeasureValues': [cpuUtilization, memoryUtilization],
        'MeasureValueType': 'MULTI',
    }

    const records = [computationalRecord];

    const params = {
        DatabaseName: constants.MEASURE_VALUE_SAMPLE_DB,
        TableName: constants.MEASURE_VALUE_SAMPLE_TABLE,
        Records: records,
        CommonAttributes: commonAttributes
    };

    const request = writeClient.writeRecords(params);

    await request.promise().then(
        (data) => {
            console.log("Write records successful");
        },
        (err) => {
            console.log("Error writing records:", err);
        }
    );
}

async function writeRecordsWithMultiMeasureValueMultipleRecordsMixture() {
    console.log("Writing records with Multi Measure Value with multiple records");
    const currentTime = Date.now().toString(); // Unix time in milliseconds

    const dimensions = [
        {'Name': 'region', 'Value': 'us-east-1'},
        {'Name': 'az', 'Value': 'az1'},
        {'Name': 'hostname', 'Value': 'host1'}
    ];

    const commonAttributes = {
        'Dimensions': dimensions,
        'Time': currentTime.toString()
    };

    const cpuUtilization = {
        'Name': 'cpu_utilization',
        'Value': '15',
        'Type': 'DOUBLE',
    };

    const memoryUtilization = {
        'Name': 'memory_utilization',
        'Value': '60',
        'Type': 'DOUBLE',
    };

    const activeCores = {
        'Name': 'active_cores',
        'Value': '4',
        'Type': 'BIGINT',
    }

    const computationalRecord = {
        'MeasureName': 'computation_utilization',
        'MeasureValues': [cpuUtilization, memoryUtilization, activeCores],
        'MeasureValueType': 'MULTI',
    }

    const records = [computationalRecord];

    const params = {
        DatabaseName: constants.MEASURE_VALUE_SAMPLE_DB,
        TableName: constants.MEASURE_VALUE_SAMPLE_TABLE,
        Records: records,
        CommonAttributes: commonAttributes
    };

    const request = writeClient.writeRecords(params);

    await request.promise().then(
        (data) => {
            console.log("Write records successful");
        },
        (err) => {
            console.log("Error writing records:", err);
        }
    );
}

module.exports = {
    writeRecordsWithMultiMeasureValueSingleRecord,
    writeRecordsWithMultiMeasureValueMultipleRecordsMixture,
}

File: sample_apps_reinvent2021/dotnet/MultiValueAttributesExample.cs
using System;
using System.IO;
using System.Collections.Generic;
using Amazon.TimestreamWrite;
using Amazon.TimestreamWrite.Model;
using System.Threading.Tasks;

namespace TimestreamDotNetSample
{
    static class MultiMeasureValueConstants
    {
        public const string MultiMeasureValueSampleDb = "multiMeasureValueSampleDb";
        public const string MultiMeasureValueSampleTable = "multiMeasureValueSampleTable";
    }

    public class MultiValueAttributesExample
    {
        private readonly AmazonTimestreamWriteClient writeClient;

        public MultiValueAttributesExample(AmazonTimestreamWriteClient writeClient)
        {
            this.writeClient = writeClient;
        }

        public async Task WriteRecordsMultiMeasureValueSingleRecord()
        {
            Console.WriteLine("Writing records with multi value attributes");

            DateTimeOffset now = DateTimeOffset.UtcNow;
            string currentTimeString = (now.ToUnixTimeMilliseconds()).ToString();

            List<Dimension> dimensions = new List<Dimension>{
                new Dimension { Name = "region", Value = "us-east-1" },
                new Dimension { Name = "az", Value = "az1" },
                new Dimension { Name = "hostname", Value = "host1" }
            };

            var commonAttributes = new Record
            {
                Dimensions = dimensions,
                Time = currentTimeString
            };

            var cpuUtilization = new MeasureValue
            {
                Name = "cpu_utilization",
                Value = "13.6",
                Type = "DOUBLE"
            };

            var memoryUtilization = new MeasureValue
            {
                Name = "memory_utilization",
                Value = "40",
                Type = "DOUBLE"
            };

            var computationalRecord = new Record
            {
                MeasureName = "cpu_memory",
                MeasureValues = new List<MeasureValue> {cpuUtilization, memoryUtilization},
                MeasureValueType = "MULTI"
            };


            List<Record> records = new List<Record>();
            records.Add(computationalRecord);

            try
            {
                var writeRecordsRequest = new WriteRecordsRequest
                {
                    DatabaseName = MultiMeasureValueConstants.MultiMeasureValueSampleDb,
                    TableName = MultiMeasureValueConstants.MultiMeasureValueSampleTable,
                    Records = records,
                    CommonAttributes = commonAttributes
                };
                WriteRecordsResponse response = await writeClient.WriteRecordsAsync(writeRecordsRequest);
                Console.WriteLine($"Write records status code: {response.HttpStatusCode.ToString()}");
            }
            catch (Exception e)
            {
                Console.WriteLine("Write records failure:" + e.ToString());
            }
        }

        public async Task WriteRecordsMultiMeasureValueMultipleRecords()
        {
            Console.WriteLine("Writing records with multi value attributes mixture type");

            DateTimeOffset now = DateTimeOffset.UtcNow;
            string currentTimeString = (now.ToUnixTimeMilliseconds()).ToString();

            List<Dimension> dimensions = new List<Dimension>{
                new Dimension { Name = "region", Value = "us-east-1" },
                new Dimension { Name = "az", Value = "az1" },
                new Dimension { Name = "hostname", Value = "host1" }
            };

            var commonAttributes = new Record
            {
                Dimensions = dimensions,
                Time = currentTimeString
            };

            var cpuUtilization = new MeasureValue
            {
                Name = "cpu_utilization",
                Value = "13.6",
                Type = "DOUBLE"
            };

            var memoryUtilization = new MeasureValue
            {
                Name = "memory_utilization",
                Value = "40",
                Type = "DOUBLE"
            };

            var activeCores = new MeasureValue
            {
                Name = "active_cores",
                Value = "4",
                Type = "BIGINT"
            };

            var computationalRecord = new Record
            {
                MeasureName = "computational_utilization",
                MeasureValues = new List<MeasureValue> {cpuUtilization, memoryUtilization, activeCores},
                MeasureValueType = "MULTI"
            };

            var aliveRecord = new Record
            {
                MeasureName = "is_healthy",
                MeasureValue = "true",
                MeasureValueType = "BOOLEAN"
            };

            List<Record> records = new List<Record>();
            records.Add(computationalRecord);
            records.Add(aliveRecord);

            try
            {
                var writeRecordsRequest = new WriteRecordsRequest
                {
                    DatabaseName = MultiMeasureValueConstants.MultiMeasureValueSampleDb,
                    TableName = MultiMeasureValueConstants.MultiMeasureValueSampleTable,
                    Records = records,
                    CommonAttributes = commonAttributes
                };
                WriteRecordsResponse response = await writeClient.WriteRecordsAsync(writeRecordsRequest);
                Console.WriteLine($"Write records status code: {response.HttpStatusCode.ToString()}");
            }
            catch (Exception e)
            {
                Console.WriteLine("Write records failure:" + e.ToString());
            }
        }
    }
}